frontend/catalyst/compilation_pipelines.py

```diff
@@ -688,13 +755,33 @@ def get_cmain(self, *args):
         return function.get_cmain(*args)
 
     def __call__(self, *args, **kwargs):
+        static_argnums = self.compile_options.static_argnums
+
         if EvaluationContext.is_tracing():
             return self.user_function(*args, **kwargs)
 
+        # Prevent self.compiled_function from being closed if any static argument exists.
+        if static_argnums:
+            self.compiled_function = None
```

Comment:
The ``compile`` method will clean up the current ``compiled_function`` before compiling any new function. (I think it is caused by ``self.compiled_function.shared_object.close()``).  I set ``compiled_function`` to None to prevent any existing function from being closed.

Reply:
Good catch, referring to my [comment further up](https://github.com/PennyLaneAI/catalyst/pull/476#discussion_r1468156643) this would correspond to the "have all versions open in memory" approach. 

I think unsetting the property in this way to prevent a different piece code from doing a specific thing is a bit hacky (because it strongly depends on the behaviour of the other code), is there a way we could refactor this life cycle management to be cleaner?

Reply:
My first thought is to disable ``self.compiled_function.shared_object.close()`` if ``static_argnums`` exists (for the "have all versions open in memory" approach ).

If we switch to the "close and reopen" approach. Then we will not face this issue since the old shared object will be restored. (But we might need to change the ``close`` method to avoid deleting old files.)

Reply:
Going to resolve this since it's no longer relevant. Ongoing discussion in https://github.com/PennyLaneAI/catalyst/pull/476#discussion_r1469998913